internal/antewrapper/msg_fees_decorator.go

```diff
@@ -0,0 +1,240 @@
+// If fee is too low, decorator returns error and tx is rejected from mempool.
+// Note this only applies when ctx.CheckTx = true
+// If fee is high enough or not CheckTx, then call next AnteHandler
+// CONTRACT: Tx must implement FeeTx to use MsgFeesDecorator
+type MsgFeesDecorator struct {
+	msgFeeKeeper   msgfeestypes.MsgFeesKeeper
+	bankKeeper     banktypes.Keeper
+	accountKeeper  cosmosante.AccountKeeper
+	feegrantKeeper msgfeestypes.FeegrantKeeper
+}
+
+func NewMsgFeesDecorator(bankKeeper banktypes.Keeper, accountKeeper cosmosante.AccountKeeper, feegrantKeeper msgfeestypes.FeegrantKeeper, keeper msgfeestypes.MsgFeesKeeper) MsgFeesDecorator {
+	return MsgFeesDecorator{
+		keeper,
+		bankKeeper,
+		accountKeeper,
+		feegrantKeeper,
+	}
+}
+
+// AnteHandle handles msg fee checking
+// has two functions ensures,
+// 1. has enough fees to add to Mempool (this involves CheckTx)
+// 2. Makes sure enough fees are present for additional message fees
+// Let z be the Total Fees to be paid
+// Let x be the Base gas Fees to be paid
+// Let y is the additional fees to be paid per MsgType
+// then z = x + y
+// This Fee Decorator makes sure that z is >= to x + y
+func (afd MsgFeesDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	feeTx, err := getFeeTx(tx)
+
+	if err != nil {
+		return ctx, err
+	}
+
+	ctx.Logger().Debug(fmt.Sprintf("In MsgFeesDecorator %d", ctx.GasMeter().GasConsumed()))
+
+	feeCoins := feeTx.GetFee()
+	gas := feeTx.GetGas()
+	msgs := feeTx.GetMsgs()
+
+	// Compute msg additionalFees
+	additionalFees, err := CalculateAdditionalFeesToBePaid(ctx, afd.msgFeeKeeper, msgs...)
+	if err != nil {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, err.Error())
+	}
+	if !additionalFees.IsZero() {
+		// ensure enough fees to cover mempool fee for base fee + additional fee
+		// This is exact same logic as NewMempoolFeeDecorator except it accounts for additional Fees.
+		if ctx.IsCheckTx() && !simulate {
+			errFromMempoolCalc := EnsureSufficientMempoolFees(ctx, gas, feeCoins, additionalFees)
+			if errFromMempoolCalc != nil {
+				return ctx, sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, errFromMempoolCalc.Error())
+			}
+		}
+		feePayer := feeTx.FeePayer()
+		feeGranter := feeTx.FeeGranter()
+
+		deductFeesFrom := feePayer
+
+		deductFeesFrom, errorFromFeeGrant := getFeeGranterIfExists(ctx, feeGranter, afd, feePayer, deductFeesFrom)
+		if errorFromFeeGrant != nil {
+			return ctx, errorFromFeeGrant
+		}
+
+		deductFeesFromAcc := afd.accountKeeper.GetAccount(ctx, deductFeesFrom)
+		if deductFeesFromAcc == nil {
+			err = sdkerrors.Wrapf(sdkerrors.ErrUnknownAddress, "fee payer address: %q does not exist", deductFeesFrom)
+			if err != nil {
+				return ctx, err
+			}
+		}
+
+		// get all the coin balances for the fee payer account
+		balancePerCoin := make(sdk.Coins, len(feeCoins))
```

Comment:
You can use `sdk.NewCoins()` here... this has the side benefit of ensuring the coin array is sorted which can prevent some determinism issues.

Reply:
i changed to this ```balancePerCoin := sdk.NewCoins()
		for _, fc := range feeCoins {
			balancePerCoin.Add(afd.bankKeeper.GetBalance(ctx, deductFeesFrom, fc.Denom))
		}
```  i think i was missing using the `Add` here which should make it sorted